sbi/neural_nets/estimators/categorical_net.py

```diff
@@ -21,93 +23,134 @@ class CategoricalNet(nn.Module):
-        """Return categorical log probability of categories input, given condition.
+    def log_prob(self, inputs: Tensor, context: Optional[Tensor] = None) -> Tensor:
+        r"""Return log-probability of samples.
 
         Args:
-            input: categories to evaluate.
-            condition: parameters.
+            input: Input datapoints of shape `(batch_size, *input_shape)`.
+            context: Context of shape `(batch_size, *condition_shape)`.
 
         Returns:
-            Tensor: log probs with shape (input.shape[0],)
+            Log-probabilities of shape `(batch_size, num_variables, num_categories)`.
         """
-        # Predict categorical ps and evaluate.
-        ps = self.forward(condition)
-        # Squeeze the last dimension (event dim) because `Categorical` has
-        # `event_shape=()` but our data usually has an event_shape of `(1,)`.
-        return Categorical(probs=ps).log_prob(input.squeeze(dim=-1))
-
-    def sample(self, sample_shape: torch.Size, condition: Tensor) -> Tensor:
-        """Returns samples from categorical random variable with probs predicted from
-        the neural net.
+        outputs = self.forward(inputs, context=context)
+        outputs = outputs.reshape(*inputs.shape, self.num_categories)
+        ps = self.compute_probs(outputs)
+
+        # categorical log prob
+        log_prob = torch.log(ps.gather(-1, inputs.unsqueeze(-1).long()))
+        log_prob = log_prob.squeeze(-1).sum(dim=-1)
+
+        return log_prob
+
+    def sample(self, sample_shape, context=None):
+        # Ensure sample_shape is a tuple
+        if isinstance(sample_shape, int):
+            sample_shape = (sample_shape,)
+        sample_shape = torch.Size(sample_shape)
+
+        # Calculate total number of samples
+        num_samples = int(torch.prod(torch.tensor(sample_shape)))
+
+        # Prepare context
+        if context is not None:
+            batch_dim = context.shape[0]
+            if context.ndim == 2:
+                context = context.unsqueeze(0)
+            if batch_dim == 1:
+                context = torchutils.repeat_rows(context, num_samples)
+        else:
+            context_dim = 0 if self.context_features is None else self.context_features
+            context = torch.zeros(num_samples, context_dim)
+            batch_dim = 1
 
-        Args:
-            sample_shape: number of samples to obtain.
-            condition: batch of parameters for prediction.
+        with torch.no_grad():
+            samples = torch.zeros(num_samples, batch_dim, self.num_variables)
+            print(samples.shape, context.shape)
+            for i in range(self.num_variables):
+                outputs = self.forward(samples, context)
+                outputs = outputs.reshape(*samples.shape, self.num_categories)
+                ps = self.compute_probs(outputs)
+                samples[:, :, i] = Categorical(probs=ps[:, :, i]).sample()
```

Comment:
same question as above: these samples are internally autoregressive, right? So each discrete variable is sampled given the upstream discrete variables?

Reply:
I am just confused because I would have expected that we for each iteration we need pass the so far sampled discrete samples as `context`; but this seems to be happening implicitly in the MADE?

Reply:
Now I see it: in line 148 you are updating `samples` with the new samples of the current `i`. It probably boils down to the same thing, but you could also update all sofar sampled `samples`, i.e., 

```
amples[:, :, :(i+1)] = Categorical(probs=ps[:, :, :(i+1)]).sample()
```
?

Reply:
I dont think it will make a difference whether the prev dims are updated since the dims are autoregressive, but I can make that change

Reply:
no need, it was just for my understanding. thanks